### DISCUSSION TITLE
对bool类型查询结果用三元表达式隐射时的一个小问题

### DISCUSSION THREAD

**Discussion by sy-ing** [dotnetcore/FreeSql#770] [General]
State = a.State == null ? false : (bool)a.State  这么写是正确的
State = a.State == null ? false : Convert.ToBoolean(a.State),  这样写生成语句报错

**Comment by luoyunchong** [comment]
什么数据库，我用mysql，第一种写反而是错误的。

**Reply by sy-ing** [reply]
sqlserver

**Comment by 2881099** [comment]
Convert.ToBoolean 解析时会翻译成 cast ( xx as

**Comment by luoyunchong** [comment]
sql server local db test
```cs
 var testDto11 = select.LeftJoin(a => a.Type.Guid == a.TypeGuid).Limit(10).ToList(a => new TestDto
            {
                id = a.Id,
                name = a.Title,
                State = a.State == null ? false : (Boolean)a.State
            });
```
这是正常的
```sql
 SELECT TOP 10 a__Type.[ParentId] as1, a.[Id] as2, a.[Title] as3, (case when a.[State]  IS  NULL then 0 else a.[State] end) as4 
FROM [tb_topic22] a 
LEFT JOIN [TestTypeInfo] a__Type ON a__Type.[Guid] = a.[TypeGuid]

```


换成Convert.ToBoolean(a.State)

```cs
       var testDto12 = select.LeftJoin(a => a.Type.Guid == a.TypeGuid).Limit(10).ToList(a => new TestDto
            {
                id = a.Id,
                name = a.Title,
                State = a.State == null ? false : Convert.ToBoolean(a.State)
            });
```
生成的sql多了not in  报错 Incorrect syntax near the keyword 'not'.
```sql
SELECT TOP 10 a__Type.[ParentId] as1, a.[Id] as2, a.[Title] as3, (case when a.[State]  IS  NULL then 0 else (cast(a.[State] as varchar) not in ('0','false')) end) as4 
FROM [tb_topic22] a 
LEFT JOIN [TestTypeInfo] a__Type ON a__Type.[Guid] = a.[TypeGuid]
```

[https://github.com/dotnetcore/FreeSql/blob/master/Providers/FreeSql.Provider.SqlServer/SqlServerExpression.cs#L32](https://github.com/dotnetcore/FreeSql/blob/master/Providers/FreeSql.Provider.SqlServer/SqlServerExpression.cs#L32)

![image](https://user-images.githubusercontent.com/18613266/119002790-bfa3b080-b9bf-11eb-83a7-b79542179e75.png)


在sql server 中，应该是else 不支持 表达式？

**Reply by 2881099** [reply]
是的，顾此失彼

**Comment by sy-ing** [comment]
在sqlserver下
 Enable2 = i.Enable == null ? "0" : Convert.ToBoolean(i.Enable) == true ? "1" : "0",//正确
 Enable2 = (bool)i.Enable == true ? "1" : "0",//错误  
和上面作为结果不同 作为条件 反而需要Convert.ToBoolean
